### DISCUSSION TITLE
Dataframe relations

### DISCUSSION THREAD

**Discussion by PetrSomekNxt** [dflib/dflib#319] [General]
Hello,
we are beginning to use dflib in our project. In our case we have class which is responsible for downloading data from some sources and this class should return this data in dataframe structure. But there is a problem - we are downloading multiple datas - for example: customers and orders and we need to write them into dataframe with the relation between each other.

Is there any way how to store these complex data into dflib dataframe structure? 

My first idea was to have dataframe with customers, this dataframe has column: 'orders', and here is written dataframe with orders for specified customer. So we have dataframe nested into dataframe, but I'm afraid about memory and performance. 
Another idea was that class with returns 2 dataframes: orders and customers. And customers dataframe has column where is written list of index (or id) of orders.

But I'd like to ask, is there any support for handling relations between dataframes?
Thanks

**Comment by andrus** [comment]
Hi! DataFrames are both Java objects and "tables", so relationship modeling approaches from the OO and the relational worlds would both work:

> My first idea was to have dataframe with customers, this dataframe has column: 'orders', and here is written dataframe with orders for specified customer. So we have dataframe nested into dataframe, but I'm afraid about memory and performance.

This is an option. It should not use more memory than any alternatives. As for performance, it depends on how you are creating and useing the resulting DataFrame

> Another idea was that class with returns 2 dataframes: orders and customers. And customers dataframe has column where is written list of index (or id) of orders.

This is a classic "one to many relationship" scenario from the relational database world. So how about doing the opposite, and modeling it the way you would model it in a database? I.e., instead of `order_ids` column in the customer DataFrame, you'd add `customer_id` column in the order DataFrame. Then you can join the two when you need to locate the orders.

Also, here is the third idea - store this data in a fully denormalized single DataFrame (essentially an outer join between customers and orders). This will require more memory then the previous solutions, but such a "flat" structure is usually the easiest to work with (for analytics, etc.) :

```
cust_id  | order_id | other columns...
-------- | -------- | -----
1        | 1        |
1        | 2        |
1        | 3        |
2        | 4        |
2        | 5        |
3        |null      |    # a customer with no orders
```

Which option is a better fit depends on how you are loading and consuming this data. If you can provide more details, maybe we can fine-tune the answer.
